File: microsoft-account-creator/telegram_bot.js
const TelegramBot = require('node-telegram-bot-api');
const fs = require('fs');
const path = require('path');

// Bot configuration
const BOT_TOKEN = process.env.TELEGRAM_BOT_TOKEN;
const ADMIN_IDS = (process.env.ADMIN_IDS || '').split(',').map(id => id.trim()).filter(Boolean);

if (!BOT_TOKEN) {
  console.error('ERROR: TELEGRAM_BOT_TOKEN environment variable is required');
  process.exit(1);
}

const bot = new TelegramBot(BOT_TOKEN, { polling: true });

// Paths
const configPath = path.join(__dirname, 'src', 'config.js');
const accountsPath = path.join(__dirname, 'accounts.txt');
const proxiesPath = path.join(__dirname, 'proxies.txt');

// State
let isGenerating = false;
let generatorProcess = null;

// Check admin
function isAdmin(userId) {
  if (ADMIN_IDS.length === 0) return true;
  return ADMIN_IDS.includes(userId.toString());
}

// Load/Save config
function loadConfig() {
  delete require.cache[require.resolve(configPath)];
  return require(configPath);
}

function saveConfig(config) {
  const content = `module.exports = ${JSON.stringify(config, null, 2).replace(/"([^"]+)":/g, '$1:')};`;
  fs.writeFileSync(configPath, content);
}

// Main menu
function getMainMenu() {
  return {
    reply_markup: {
      inline_keyboard: [
        [
          { text: '🚀 Generate', callback_data: 'menu_generate' },
          { text: '📊 Status', callback_data: 'status' }
        ],
        [
          { text: '📋 Accounts', callback_data: 'accounts' },
          { text: '⚙️ Settings', callback_data: 'menu_settings' }
        ],
        [
          { text: '🌐 Proxy', callback_data: 'menu_proxy' },
          { text: '❓ Help', callback_data: 'help' }
        ]
      ]
    }
  };
}

// /start
bot.onText(/\/start/, (msg) => {
  if (!isAdmin(msg.from.id)) return bot.sendMessage(msg.chat.id, '❌ Unauthorized');

  const config = loadConfig();
  bot.sendMessage(msg.chat.id, 
    `🔐 *Microsoft Account Creator*\n\n` +
    `🎯 Captcha: \`2Captcha\`\n` +
    `🌐 Proxy: \`${config.USE_PROXY ? 'Webshare.io' : 'OFF'}\`\n` +
    `📧 Domain: \`${config.EMAIL_DOMAIN}\`\n` +
    `🔄 Recovery: \`${config.ADD_RECOVERY_EMAIL ? 'ON' : 'OFF'}\``,
    { parse_mode: 'Markdown', ...getMainMenu() }
  );
});

// Callback handler
bot.on('callback_query', async (query) => {
  if (!isAdmin(query.from.id)) {
    return bot.answerCallbackQuery(query.id, { text: '❌ Unauthorized' });
  }

  const chatId = query.message.chat.id;
  const msgId = query.message.message_id;
  const config = loadConfig();

  switch (query.data) {
    case 'menu_generate':
      await bot.editMessageText('🚀 *Generate Accounts*\n\nPilih jumlah:', {
        chat_id: chatId, message_id: msgId, parse_mode: 'Markdown',
        reply_markup: {
          inline_keyboard: [
            [
              { text: '1️⃣', callback_data: 'gen_1' },
              { text: '5️⃣', callback_data: 'gen_5' },
              { text: '🔟', callback_data: 'gen_10' }
            ],
            [
              { text: '25', callback_data: 'gen_25' },
              { text: '50', callback_data: 'gen_50' },
              { text: '100', callback_data: 'gen_100' }
            ],
            [{ text: '⬅️ Back', callback_data: 'back_main' }]
          ]
        }
      });
      break;

    case 'menu_settings':
      await bot.editMessageText(
        `⚙️ *Settings*\n\n` +
        `📧 Domain: \`${config.EMAIL_DOMAIN}\`\n` +
        `🔄 Recovery: \`${config.ADD_RECOVERY_EMAIL ? 'ON' : 'OFF'}\`\n` +
        `🖥️ Headless: \`${config.HEADLESS ? 'ON' : 'OFF'}\``,
        {
          chat_id: chatId, message_id: msgId, parse_mode: 'Markdown',
          reply_markup: {
            inline_keyboard: [
              [{ text: `📧 Domain: ${config.EMAIL_DOMAIN}`, callback_data: 'toggle_domain' }],
              [{ text: `🔄 Recovery: ${config.ADD_RECOVERY_EMAIL ? '✅' : '❌'}`, callback_data: 'toggle_recovery' }],
              [{ text: `🖥️ Headless: ${config.HEADLESS ? '✅' : '❌'}`, callback_data: 'toggle_headless' }],
              [{ text: '⬅️ Back', callback_data: 'back_main' }]
            ]
          }
        }
      );
      break;

    case 'menu_proxy':
      await bot.editMessageText(
        `🌐 *Webshare.io Rotating Proxy*\n\n` +
        `Status: \`${config.USE_PROXY ? 'ON' : 'OFF'}\`\n` +
        `Host: \`${config.PROXY_HOST || 'p.webshare.io'}\`\n` +
        `Port: \`${config.PROXY_PORT || '80'}\`\n` +
        `User: \`${config.PROXY_USERNAME || 'Not set'}\`\n\n` +
        `Set proxy dengan:\n\`/setproxy user:pass@host:port\`\n` +
        `Contoh: \`/setproxy nyrtjgvw-rotate:xxx@p.webshare.io:80\``,
        {
          chat_id: chatId, message_id: msgId, parse_mode: 'Markdown',
          reply_markup: {
            inline_keyboard: [
              [{ text: `🌐 Proxy: ${config.USE_PROXY ? '✅ ON' : '❌ OFF'}`, callback_data: 'toggle_proxy' }],
              [{ text: '⬅️ Back', callback_data: 'back_main' }]
            ]
          }
        }
      );
      break;

    case 'toggle_domain':
      config.EMAIL_DOMAIN = config.EMAIL_DOMAIN === '@outlook.com' ? '@hotmail.com' : '@outlook.com';
      saveConfig(config);
      await bot.answerCallbackQuery(query.id, { text: `Domain: ${config.EMAIL_DOMAIN}` });
      bot.emit('callback_query', { ...query, data: 'menu_settings' });
      break;

    case 'toggle_recovery':
      config.ADD_RECOVERY_EMAIL = !config.ADD_RECOVERY_EMAIL;
      saveConfig(config);
      await bot.answerCallbackQuery(query.id, { text: `Recovery: ${config.ADD_RECOVERY_EMAIL ? 'ON' : 'OFF'}` });
      bot.emit('callback_query', { ...query, data: 'menu_settings' });
      break;

    case 'toggle_headless':
      config.HEADLESS = !config.HEADLESS;
      saveConfig(config);
      await bot.answerCallbackQuery(query.id, { text: `Headless: ${config.HEADLESS ? 'ON' : 'OFF'}` });
      bot.emit('callback_query', { ...query, data: 'menu_settings' });
      break;

    case 'toggle_proxy':
      config.USE_PROXY = !config.USE_PROXY;
      saveConfig(config);
      await bot.answerCallbackQuery(query.id, { text: `Proxy: ${config.USE_PROXY ? 'ON' : 'OFF'}` });
      bot.emit('callback_query', { ...query, data: 'menu_proxy' });
      break;

    case 'gen_1': case 'gen_5': case 'gen_10': 
    case 'gen_25': case 'gen_50': case 'gen_100':
      const count = parseInt(query.data.split('_')[1]);
      await startGeneration(chatId, msgId, count);
      break;

    case 'status':
      let accountCount = 0;
      try {
        const accounts = fs.readFileSync(accountsPath, 'utf8');
        accountCount = accounts.split('\n').filter(l => l.includes(':')).length;
      } catch (e) {}
      
      await bot.editMessageText(
        `📊 *Status*\n\n` +
        `🔄 Running: \`${isGenerating ? 'YES' : 'NO'}\`\n` +
        `📧 Total: \`${accountCount} accounts\`\n` +
        `🎯 2Captcha Key: \`${config.CAPTCHA_API_KEY ? 'Set ✅' : 'Not set ❌'}\`\n` +
        `🌐 Proxy: \`${config.USE_PROXY ? 'ON' : 'OFF'}\``,
        { chat_id: chatId, message_id: msgId, parse_mode: 'Markdown', ...getMainMenu() }
      );
      break;

    case 'accounts':
      try {
        const accounts = fs.readFileSync(accountsPath, 'utf8');
        const lines = accounts.split('\n').filter(l => l.includes(':'));
        const last10 = lines.slice(-10);
        
        await bot.editMessageText(
          last10.length > 0
            ? `📋 *Last 10 Accounts*\n\n\`\`\`\n${last10.join('\n')}\n\`\`\`\nTotal: ${lines.length}`
            : '📋 *No accounts yet*',
          {
            chat_id: chatId, message_id: msgId, parse_mode: 'Markdown',
            reply_markup: {
              inline_keyboard: [
                [{ text: '📥 Download', callback_data: 'download' }],
                [{ text: '⬅️ Back', callback_data: 'back_main' }]
              ]
            }
          }
        );
      } catch (e) {
        await bot.editMessageText('📋 *No accounts file*', 
          { chat_id: chatId, message_id: msgId, parse_mode: 'Markdown', ...getMainMenu() });
      }
      break;

    case 'download':
      try {
        await bot.sendDocument(chatId, accountsPath, { caption: '📥 Generated Accounts' });
      } catch (e) {
        await bot.sendMessage(chatId, '❌ No accounts file');
      }
      break;

    case 'stop':
      if (generatorProcess) {
        generatorProcess.kill();
        generatorProcess = null;
      }
      isGenerating = false;
      await bot.editMessageText('🛑 *Stopped*', 
        { chat_id: chatId, message_id: msgId, parse_mode: 'Markdown', ...getMainMenu() });
      break;

    case 'help':
      await bot.editMessageText(
        `❓ *Help*\n\n` +
        `*Commands:*\n` +
        `/start - Main menu\n` +
        `/setapikey KEY - Set 2Captcha API key\n` +
        `/setproxy endpoint:port - Set rotating proxy\n` +
        `/balance - Check 2Captcha balance\n\n` +
        `*Webshare Rotating Proxy:*\n` +
        `1. Whitelist VPS IP di Webshare\n` +
        `2. Set ke IP Authentication\n` +
        `3. Set ke Rotating Proxy Endpoint\n` +
        `4. Copy endpoint: \`p.webshare.io:80\``,
        { chat_id: chatId, message_id: msgId, parse_mode: 'Markdown', ...getMainMenu() }
      );
      break;

    case 'back_main':
      const cfg = loadConfig();
      await bot.editMessageText(
        `🔐 *Microsoft Account Creator*\n\n` +
        `🎯 Captcha: \`2Captcha\`\n` +
        `🌐 Proxy: \`${cfg.USE_PROXY ? 'ON' : 'OFF'}\`\n` +
        `📧 Domain: \`${cfg.EMAIL_DOMAIN}\``,
        { chat_id: chatId, message_id: msgId, parse_mode: 'Markdown', ...getMainMenu() }
      );
      break;
  }
  
  await bot.answerCallbackQuery(query.id);
});

// Start generation
async function startGeneration(chatId, msgId, count) {
  if (isGenerating) {
    return bot.editMessageText('⚠️ Already running!', {
      chat_id: chatId, message_id: msgId,
      reply_markup: { inline_keyboard: [[{ text: '🛑 Stop', callback_data: 'stop' }]] }
    });
  }

  const config = loadConfig();
  if (!config.CAPTCHA_API_KEY) {
    return bot.editMessageText('❌ Set 2Captcha API key first!\n\n`/setapikey YOUR_KEY`', 
      { chat_id: chatId, message_id: msgId, parse_mode: 'Markdown', ...getMainMenu() });
  }

  isGenerating = true;
  
  await bot.editMessageText(`🚀 *Generating ${count} accounts...*\n\n⏳ Please wait...`, {
    chat_id: chatId, message_id: msgId, parse_mode: 'Markdown',
    reply_markup: { inline_keyboard: [[{ text: '🛑 Stop', callback_data: 'stop' }]] }
  });

  const { spawn } = require('child_process');
  generatorProcess = spawn('node', ['src/index.js', count.toString()], {
    cwd: __dirname,
    env: process.env
  });

  let output = '';
  
  generatorProcess.stdout.on('data', (data) => {
    output += data.toString();
    console.log(data.toString());
  });

  generatorProcess.stderr.on('data', (data) => console.error(data.toString()));

  generatorProcess.on('close', async () => {
    isGenerating = false;
    generatorProcess = null;
    const successCount = (output.match(/SUCCESS:/g) || []).length;
    
    await bot.sendMessage(chatId,
      `✅ *Done!*\n\n📊 Generated: ${successCount}/${count}`,
      { parse_mode: 'Markdown', ...getMainMenu() }
    );
  });
}

// /setapikey
bot.onText(/\/setapikey (.+)/, (msg, match) => {
  if (!isAdmin(msg.from.id)) return;
  
  const config = loadConfig();
  config.CAPTCHA_API_KEY = match[1].trim();
  config.CAPTCHA_PROVIDER = 'twocaptcha';
  saveConfig(config);
  
  bot.sendMessage(msg.chat.id, `✅ 2Captcha API key set!\n\nKey: \`${match[1].trim().substring(0, 8)}....\``, 
    { parse_mode: 'Markdown' });
});

// /setproxy (Format: user:pass@host:port)
bot.onText(/\/setproxy (.+)/, (msg, match) => {
  if (!isAdmin(msg.from.id)) return;
  
  const proxy = match[1].trim();
  const config = loadConfig();
  
  if (proxy.includes('@')) {
    // Format: user:pass@host:port
    const [auth, addr] = proxy.split('@');
    const [user, pass] = auth.split(':');
    const [host, port] = addr.split(':');
    config.PROXY_USERNAME = user;
    config.PROXY_PASSWORD = pass;
    config.PROXY_HOST = host;
    config.PROXY_PORT = port || '80';
  } else {
    // Format: host:port (IP auth, no credentials)
    const [host, port] = proxy.split(':');
    config.PROXY_HOST = host;
    config.PROXY_PORT = port || '80';
    config.PROXY_USERNAME = '';
    config.PROXY_PASSWORD = '';
  }
  
  config.USE_PROXY = true;
  saveConfig(config);
  
  bot.sendMessage(msg.chat.id, 
    `✅ Rotating proxy configured!\n\n` +
    `🌐 Host: \`${config.PROXY_HOST}:${config.PROXY_PORT}\`\n` +
    `👤 User: \`${config.PROXY_USERNAME || 'IP Auth'}\``,
    { parse_mode: 'Markdown' }
  );
});

// /balance - Check 2captcha balance
bot.onText(/\/balance/, async (msg) => {
  if (!isAdmin(msg.from.id)) return;
  
  const config = loadConfig();
  if (!config.CAPTCHA_API_KEY) {
    return bot.sendMessage(msg.chat.id, '❌ Set API key first: `/setapikey KEY`', { parse_mode: 'Markdown' });
  }
  
  const https = require('https');
  const url = `https://2captcha.com/res.php?key=${config.CAPTCHA_API_KEY}&action=getbalance&json=1`;
  
  https.get(url, (res) => {
    let data = '';
    res.on('data', chunk => data += chunk);
    res.on('end', () => {
      try {
        const result = JSON.parse(data);
        if (result.status === 1) {
          bot.sendMessage(msg.chat.id, `💰 *2Captcha Balance*\n\n$${result.request}`, { parse_mode: 'Markdown' });
        } else {
          bot.sendMessage(msg.chat.id, `❌ Error: ${result.request}`);
        }
      } catch (e) {
        bot.sendMessage(msg.chat.id, '❌ Failed to check balance');
      }
    });
  });
});

console.log('🤖 Bot started!');
console.log(`Admin IDs: ${ADMIN_IDS.length > 0 ? ADMIN_IDS.join(', ') : 'Open access'}`);

File: outlook_gen/telegram_bot.py
#!/usr/bin/env python3
"""
Telegram Bot untuk kontrol OutlookGen
"""

import os
import json
import asyncio
import subprocess
from datetime import datetime
from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import Application, CommandHandler, CallbackQueryHandler, MessageHandler, filters, ContextTypes

# ================== KONFIGURASI ==================
BOT_TOKEN = os.environ.get("TELEGRAM_BOT_TOKEN", "YOUR_BOT_TOKEN_HERE")
ADMIN_IDS = [int(x) for x in os.environ.get("ADMIN_IDS", "").split(",") if x]  # Comma-separated admin IDs
WORK_DIR = os.path.dirname(os.path.abspath(__file__))
# =================================================

# Global state
generator_process = None
generated_count = 0
is_running = False

def is_admin(user_id: int) -> bool:
    """Check if user is admin"""
    if not ADMIN_IDS:
        return True  # No restriction if no admin IDs configured
    return user_id in ADMIN_IDS

def load_config() -> dict:
    """Load config.json"""
    config_path = os.path.join(WORK_DIR, "config.json")
    with open(config_path, 'r') as f:
        return json.load(f)

def save_config(config: dict):
    """Save config.json"""
    config_path = os.path.join(WORK_DIR, "config.json")
    with open(config_path, 'w') as f:
        json.dump(config, f, indent=2)

def get_accounts() -> list:
    """Get generated accounts"""
    accounts_path = os.path.join(WORK_DIR, "account.txt")
    if not os.path.exists(accounts_path):
        return []
    with open(accounts_path, 'r') as f:
        lines = [l.strip() for l in f.readlines() if l.strip() and not l.startswith('#')]
    return lines

def get_proxies() -> list:
    """Get proxy list"""
    proxy_path = os.path.join(WORK_DIR, "proxy.txt")
    if not os.path.exists(proxy_path):
        return []
    with open(proxy_path, 'r') as f:
        lines = [l.strip() for l in f.readlines() if l.strip()]
    return lines

def save_proxies(proxies: list):
    """Save proxy list"""
    proxy_path = os.path.join(WORK_DIR, "proxy.txt")
    with open(proxy_path, 'w') as f:
        f.write('\n'.join(proxies))

# ================== HANDLERS ==================

async def start(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Start command - show main menu"""
    user_id = update.effective_user.id
    if not is_admin(user_id):
        await update.message.reply_text("⛔ Unauthorized")
        return

    keyboard = [
        [InlineKeyboardButton("🚀 Generate Email", callback_data="generate")],
        [InlineKeyboardButton("📊 Status", callback_data="status"),
         InlineKeyboardButton("📧 Accounts", callback_data="accounts")],
        [InlineKeyboardButton("⚙️ Config", callback_data="config"),
         InlineKeyboardButton("🔌 Proxy", callback_data="proxy")],
        [InlineKeyboardButton("🛑 Stop Generator", callback_data="stop")]
    ]
    reply_markup = InlineKeyboardMarkup(keyboard)
    
    await update.message.reply_text(
        "🎯 *OutlookGen Telegram Bot*\n\n"
        "Pilih menu di bawah:",
        reply_markup=reply_markup,
        parse_mode="Markdown"
    )

async def button_callback(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Handle button callbacks"""
    query = update.callback_query
    user_id = query.from_user.id
    
    if not is_admin(user_id):
        await query.answer("⛔ Unauthorized", show_alert=True)
        return
    
    await query.answer()
    data = query.data

    if data == "generate":
        await show_generate_menu(query, context)
    elif data == "status":
        await show_status(query, context)
    elif data == "accounts":
        await show_accounts(query, context)
    elif data == "config":
        await show_config(query, context)
    elif data == "proxy":
        await show_proxy(query, context)
    elif data == "stop":
        await stop_generator(query, context)
    elif data.startswith("gen_"):
        count = int(data.split("_")[1])
        await start_generator(query, context, count)
    elif data == "back":
        await back_to_menu(query, context)

async def show_generate_menu(query, context):
    """Show generate options"""
    keyboard = [
        [InlineKeyboardButton("1 Email", callback_data="gen_1"),
         InlineKeyboardButton("5 Email", callback_data="gen_5")],
        [InlineKeyboardButton("10 Email", callback_data="gen_10"),
         InlineKeyboardButton("25 Email", callback_data="gen_25")],
        [InlineKeyboardButton("50 Email", callback_data="gen_50"),
         InlineKeyboardButton("100 Email", callback_data="gen_100")],
        [InlineKeyboardButton("⬅️ Back", callback_data="back")]
    ]
    reply_markup = InlineKeyboardMarkup(keyboard)
    
    await query.edit_message_text(
        "🚀 *Generate Email*\n\n"
        "Pilih jumlah email yang ingin digenerate:",
        reply_markup=reply_markup,
        parse_mode="Markdown"
    )

async def start_generator(query, context, count: int):
    """Start the generator"""
    global generator_process, is_running
    
    if is_running:
        await query.edit_message_text(
            "⚠️ Generator sudah berjalan!\n"
            "Stop dulu sebelum memulai yang baru."
        )
        return
    
    config = load_config()
    
    # Validate config
    if config['Captcha']['api_key'] == "":
        await query.edit_message_text("❌ Captcha API key belum diset!")
        return
    
    if config['Captcha']['providers'] == "anycaptcha/twocaptcha":
        await query.edit_message_text("❌ Pilih provider captcha (anycaptcha atau twocaptcha)!")
        return
    
    if config['EmailInfo']['Domain'] == "@hotmail.com/@outlook.com":
        await query.edit_message_text("❌ Pilih domain email (@hotmail.com atau @outlook.com)!")
        return
    
    proxies = get_proxies()
    if not proxies:
        await query.edit_message_text("❌ Tidak ada proxy! Tambahkan proxy dulu.")
        return
    
    is_running = True
    await query.edit_message_text(
        f"🚀 *Starting Generator*\n\n"
        f"Target: {count} email\n"
        f"Provider: {config['Captcha']['providers']}\n"
        f"Domain: {config['EmailInfo']['Domain']}\n"
        f"Proxy: {len(proxies)} available\n\n"
        f"⏳ Generating...",
        parse_mode="Markdown"
    )
    
    # Run generator in background
    # Note: This is simplified - in production you'd want better process management
    try:
        cmd = f"cd {WORK_DIR} && python3 main.py"
        generator_process = subprocess.Popen(
            cmd,
            shell=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE
        )
        
        # Monitor for a bit
        await asyncio.sleep(5)
        
        if generator_process.poll() is not None:
            # Process ended
            stdout, stderr = generator_process.communicate()
            is_running = False
            await query.edit_message_text(
                f"❌ Generator stopped unexpectedly\n\n"
                f"Error: {stderr.decode()[:500]}"
            )
        else:
            keyboard = [[InlineKeyboardButton("🛑 Stop", callback_data="stop"),
                        InlineKeyboardButton("📊 Status", callback_data="status")]]
            await query.edit_message_text(
                f"✅ Generator berjalan!\n\n"
                f"Gunakan /status untuk melihat progress.",
                reply_markup=InlineKeyboardMarkup(keyboard)
            )
    except Exception as e:
        is_running = False
        await query.edit_message_text(f"❌ Error: {str(e)}")

async def stop_generator(query, context):
    """Stop the generator"""
    global generator_process, is_running
    
    if not is_running:
        await query.edit_message_text("ℹ️ Generator tidak sedang berjalan.")
        return
    
    if generator_process:
        generator_process.terminate()
        generator_process = None
    
    is_running = False
    await query.edit_message_text("🛑 Generator dihentikan.")

async def show_status(query, context):
    """Show generator status"""
    global is_running
    
    accounts = get_accounts()
    proxies = get_proxies()
    config = load_config()
    
    status = "🟢 Running" if is_running else "🔴 Stopped"
    
    text = (
        f"📊 *Status Generator*\n\n"
        f"Status: {status}\n"
        f"Generated: {len(accounts)} accounts\n"
        f"Proxies: {len(proxies)} available\n"
        f"Provider: {config['Captcha']['providers']}\n"
        f"Domain: {config['EmailInfo']['Domain']}\n"
    )
    
    keyboard = [[InlineKeyboardButton("⬅️ Back", callback_data="back")]]
    await query.edit_message_text(text, reply_markup=InlineKeyboardMarkup(keyboard), parse_mode="Markdown")

async def show_accounts(query, context):
    """Show generated accounts"""
    accounts = get_accounts()
    
    if not accounts:
        text = "📧 *Generated Accounts*\n\nBelum ada akun yang digenerate."
    else:
        # Show last 10
        recent = accounts[-10:]
        text = f"📧 *Generated Accounts* ({len(accounts)} total)\n\n"
        text += "Last 10:\n"
        for acc in recent:
            text += f"• `{acc}`\n"
    
    keyboard = [[InlineKeyboardButton("⬅️ Back", callback_data="back")]]
    await query.edit_message_text(text, reply_markup=InlineKeyboardMarkup(keyboard), parse_mode="Markdown")

async def show_config(query, context):
    """Show current config"""
    config = load_config()
    
    text = (
        f"⚙️ *Configuration*\n\n"
        f"*Captcha:*\n"
        f"• Provider: `{config['Captcha']['providers']}`\n"
        f"• API Key: `{'*****' if config['Captcha']['api_key'] else 'NOT SET'}`\n\n"
        f"*Email:*\n"
        f"• Domain: `{config['EmailInfo']['Domain']}`\n"
        f"• Password Length: `{config['EmailInfo']['PasswordLength']}`\n\n"
        f"Untuk edit config, gunakan:\n"
        f"`/setprovider <anycaptcha/twocaptcha>`\n"
        f"`/setapikey <key>`\n"
        f"`/setdomain <@hotmail.com/@outlook.com>`"
    )
    
    keyboard = [[InlineKeyboardButton("⬅️ Back", callback_data="back")]]
    await query.edit_message_text(text, reply_markup=InlineKeyboardMarkup(keyboard), parse_mode="Markdown")

async def show_proxy(query, context):
    """Show proxy list"""
    proxies = get_proxies()
    
    if not proxies:
        text = "🔌 *Proxy List*\n\nTidak ada proxy."
    else:
        text = f"🔌 *Proxy List* ({len(proxies)} total)\n\n"
        for i, p in enumerate(proxies[:10], 1):
            # Mask proxy for security
            masked = p[:20] + "..." if len(p) > 20 else p
            text += f"{i}. `{masked}`\n"
        if len(proxies) > 10:
            text += f"\n... dan {len(proxies) - 10} lainnya"
    
    text += "\n\nUntuk tambah proxy:\n`/addproxy ip:port`\n`/clearproxy` - hapus semua"
    
    keyboard = [[InlineKeyboardButton("⬅️ Back", callback_data="back")]]
    await query.edit_message_text(text, reply_markup=InlineKeyboardMarkup(keyboard), parse_mode="Markdown")

async def back_to_menu(query, context):
    """Back to main menu"""
    keyboard = [
        [InlineKeyboardButton("🚀 Generate Email", callback_data="generate")],
        [InlineKeyboardButton("📊 Status", callback_data="status"),
         InlineKeyboardButton("📧 Accounts", callback_data="accounts")],
        [InlineKeyboardButton("⚙️ Config", callback_data="config"),
         InlineKeyboardButton("🔌 Proxy", callback_data="proxy")],
        [InlineKeyboardButton("🛑 Stop Generator", callback_data="stop")]
    ]
    reply_markup = InlineKeyboardMarkup(keyboard)
    
    await query.edit_message_text(
        "🎯 *OutlookGen Telegram Bot*\n\n"
        "Pilih menu di bawah:",
        reply_markup=reply_markup,
        parse_mode="Markdown"
    )

# ================== CONFIG COMMANDS ==================

async def set_provider(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Set captcha provider"""
    user_id = update.effective_user.id
    if not is_admin(user_id):
        return
    
    if not context.args:
        await update.message.reply_text("Usage: /setprovider <anycaptcha/twocaptcha>")
        return
    
    provider = context.args[0].lower()
    if provider not in ['anycaptcha', 'twocaptcha']:
        await update.message.reply_text("❌ Provider harus 'anycaptcha' atau 'twocaptcha'")
        return
    
    config = load_config()
    config['Captcha']['providers'] = provider
    save_config(config)
    await update.message.reply_text(f"✅ Provider diset ke: {provider}")

async def set_apikey(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Set captcha API key"""
    user_id = update.effective_user.id
    if not is_admin(user_id):
        return
    
    if not context.args:
        await update.message.reply_text("Usage: /setapikey <api_key>")
        return
    
    api_key = context.args[0]
    config = load_config()
    config['Captcha']['api_key'] = api_key
    save_config(config)
    await update.message.reply_text("✅ API key berhasil diset!")
    
    # Delete the message containing API key for security
    try:
        await update.message.delete()
    except:
        pass

async def set_domain(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Set email domain"""
    user_id = update.effective_user.id
    if not is_admin(user_id):
        return
    
    if not context.args:
        await update.message.reply_text("Usage: /setdomain <@hotmail.com/@outlook.com>")
        return
    
    domain = context.args[0]
    if domain not in ['@hotmail.com', '@outlook.com']:
        await update.message.reply_text("❌ Domain harus '@hotmail.com' atau '@outlook.com'")
        return
    
    config = load_config()
    config['EmailInfo']['Domain'] = domain
    save_config(config)
    await update.message.reply_text(f"✅ Domain diset ke: {domain}")

async def add_proxy(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Add proxy to list"""
    user_id = update.effective_user.id
    if not is_admin(user_id):
        return
    
    if not context.args:
        await update.message.reply_text("Usage: /addproxy ip:port atau user:pass@ip:port")
        return
    
    proxy = context.args[0]
    proxies = get_proxies()
    proxies.append(proxy)
    save_proxies(proxies)
    await update.message.reply_text(f"✅ Proxy ditambahkan! Total: {len(proxies)}")

async def clear_proxy(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Clear all proxies"""
    user_id = update.effective_user.id
    if not is_admin(user_id):
        return
    
    save_proxies([])
    await update.message.reply_text("✅ Semua proxy dihapus!")

async def download_accounts(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Download accounts file"""
    user_id = update.effective_user.id
    if not is_admin(user_id):
        return
    
    accounts_path = os.path.join(WORK_DIR, "account.txt")
    if not os.path.exists(accounts_path) or os.path.getsize(accounts_path) == 0:
        await update.message.reply_text("❌ Tidak ada akun yang tersimpan.")
        return
    
    await update.message.reply_document(
        document=open(accounts_path, 'rb'),
        filename=f"accounts_{datetime.now().strftime('%Y%m%d_%H%M%S')}.txt",
        caption="📧 Generated Accounts"
    )

# ================== MAIN ==================

def main():
    """Start the bot"""
    if BOT_TOKEN == "YOUR_BOT_TOKEN_HERE":
        print("❌ Please set TELEGRAM_BOT_TOKEN environment variable!")
        return
    
    application = Application.builder().token(BOT_TOKEN).build()
    
    # Commands
    application.add_handler(CommandHandler("start", start))
    application.add_handler(CommandHandler("setprovider", set_provider))
    application.add_handler(CommandHandler("setapikey", set_apikey))
    application.add_handler(CommandHandler("setdomain", set_domain))
    application.add_handler(CommandHandler("addproxy", add_proxy))
    application.add_handler(CommandHandler("clearproxy", clear_proxy))
    application.add_handler(CommandHandler("download", download_accounts))
    
    # Callbacks
    application.add_handler(CallbackQueryHandler(button_callback))
    
    print("🤖 Bot started!")
    application.run_polling(allowed_updates=Update.ALL_TYPES)

if __name__ == "__main__":
    main()
